### PR TITLE
search: update StreamSearch interface

### PR DESCRIPTION
**Pull request by stefanhengl**
This updates the `StreamSearch` interface to take a `zoektstream.Streamer`
and return an `error`. This prepares the codebase to call Zoekt's new streaming
endpoint. This PR is a pure refactor.

### CODE REVIEW

**Comment by stefanhengl** [comment]
cc @keegancsmith

File: internal/search/backend/horizontal.go [hunk 1]
```diff
@@ -82,39 +72,36 @@ func (s *HorizontalSearcher) Search(ctx context.Context, q query.Q, opts *zoekt.
...
 // AggregateStreamSearch aggregates the stream events into a single batch
 // result.
-func AggregateStreamSearch(ctx context.Context, streamSearch func(context.Context, query.Q, *zoekt.SearchOptions) <-chan StreamSearchEvent, q query.Q, opts *zoekt.SearchOptions) (*zoekt.SearchResult, error) {
+func AggregateStreamSearch(ctx context.Context, streamSearch func(context.Context, query.Q, *zoekt.SearchOptions, ZoektStreamer) error, q query.Q, opts *zoekt.SearchOptions) (*zoekt.SearchResult, error) {
 	start := time.Now()
 
+	var mu sync.Mutex
 	aggregate := &zoekt.SearchResult{
 		RepoURLs:      map[string]string{},
 		LineFragments: map[string]string{},
```
Comment (stefanhengl) [inline internal/search/backend/horizontal.go:81]: Note: Check if use RepoURLs or LineFragments anywhere. If not, remove them here.

**Comment by keegancsmith** [comment]
@stefanhengl ready for review?

**Comment by stefanhengl** [comment]
Not yet, have to incorporate the zoekt update. Shouldn't take long.

File: internal/search/backend/horizontal.go (hunk 1 above)
Comment (stefanhengl) [reply internal/search/backend/horizontal.go]: I removed RepoURLs and LineFragments. Both were just used in horizontal.go and in tests.

File: cmd/frontend/graphqlbackend/zoekt.go [hunk 2]
```diff
@@ -277,7 +280,20 @@ func zoektSearch(ctx context.Context, args *search.TextParameters, repos *indexe
...
+		cancel()
+		for range events {
+		}
+	}()
+	g := errgroup.Group{}
+	g.Go(func() error {
+		defer close(events)
+		return args.Zoekt.Client.StreamSearch(ctx, finalQuery, &searchOpts, backend.ZoektStreamFunc(func(event *zoekt.SearchResult) {
+			events <- event
+		}))
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/zoekt.go:295]: this makes the change more minimal, but I wonder if we should avoid the channel and just embed the event loop below here? either way something that we can follow-up in another PR with.

File: internal/search/backend/horizontal.go [hunk 3]
```diff
@@ -25,54 +27,42 @@ type HorizontalSearcher struct {
...
-			return
-		}
+			mu.Lock()
+			defer mu.Unlock()
 
-		if r.SearchResult != nil {
-			r.SearchResult.Files = dedupper.Dedup(r.SearchResult.Files)
-		}
+			sr.Files = dedupper.Dedup(sr.Files)
+			streamer.Send(sr)
```
Comment (keegancsmith) [inline internal/search/backend/horizontal.go:59]: you are holding the lock while you send. Probably should only hold the lock while dedupping.

**Review by keegancsmith — APPROVED** [review]

File: cmd/frontend/graphqlbackend/zoekt.go (hunk 2 above)
Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/zoekt.go:295]: I actually experimented with that (see commit history) but reverted it. It gets tricky with the repo resolution and the code ended up more convoluted than it is right now. Let's tackle this in another PR.

**Pull request merged**